### PR TITLE
Setup sorcery pw auth

### PR DESCRIPTION
**Pull request by r3drighthand**
@giladshanan 

I implemented a basic setup for password authentication with Sorcery. Plan on doing a lot more (obviously) but would like a thorough review before proceeding.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,3 +1,9 @@
 class User < ApplicationRecord
   authenticates_with_sorcery!
+
+  validates :password, length: {minimum: 3}, if: -> {:new_record? || changes[:crypted_password]}
+  validates :password, confirmation: true, if: -> {:new_record? || changes[:crypted_password]}
+  validates :password_confirmation, presence: true, if: -> {:new_record? || changes || [:crypted_password]}
```
Comment (giladshanan) [inline app/models/user.rb]: this looks like a typo
```suggestion
  validates :password_confirmation, presence: true, if: -> {:new_record? || changes[:crypted_password]}
```

File: app/controllers/user_sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def create
+    @user = login(params[:email], params[:password])
+
+    if @user 
+      redirect_back_or_to(:users, notice: "Login successful")
+    else
+      flash.now[:alert] = "Login failed"
+      render action: 'new'
```
Comment (giladshanan) [inline app/controllers/user_sessions_controller.rb]: single quotes cop
```suggestion
      render action: "new"
```

File: app/views/user_sessions/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<%= form_tag user_sessions_path, :method => :post do %>
+  <div class="field">
+    <%= label_tag :email %><br />
+    <%= text_field_tag :email %>
```
Comment (giladshanan) [inline app/views/user_sessions/_form.html.erb:4]: Let's use `email_field_tag` instead: https://apidock.com/rails/ActionView/Helpers/FormTagHelper/email_field_tag

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,9 +1,20 @@
...
+
   def index
+    # 
   end
 
-  private
+  def create
+    @user = User.new(params[:user])
+    if @user.save
+      redirect_to(:users, notice: 'User was successfully created')
```
Comment (giladshanan) [inline app/controllers/users_controller.rb:11]: single quotes

File: app/controllers/user_sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = login(user_params[:email], user_params[:password])
+
+    if @user 
+      redirect_back_or_to(:users, notice: 'Login successful')
```
Comment (giladshanan) [inline app/controllers/user_sessions_controller.rb]: In general I prefer to use double quotes.

```suggestion
      redirect_back_or_to(:users, notice: "Login successful")
```

File: app/controllers/user_sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    else
+      flash.now[:alert] = 'Login failed'
```
Comment (giladshanan) [inline app/controllers/user_sessions_controller.rb]:
```suggestion
      flash.now[:alert] = "Login failed"
```

File: app/controllers/user_sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      render action: 'new'
+    end
+  end
+
+  def destroy
+    logout
+    redirect_to(:users, notice: 'You\'ve been logged out')
```
Comment (giladshanan) [inline app/controllers/user_sessions_controller.rb:21]: single quotes

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  def index
+    # 
+  end
+
+  def create
+    @user = User.new(params[:user])
+    if @user.save
+      redirect_to(:users, notice: 'User was successfully created')
+    else
+      render 'new'
```
Comment (giladshanan) [inline app/controllers/users_controller.rb:13]: single quotes

File: app/views/user_sessions/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<h1>Login</h1>
+
+<%= render 'form' %>
```
Comment (giladshanan) [inline app/views/user_sessions/new.html.erb:3]: single quotes

**Review by giladshanan — CHANGES_REQUESTED** [review]
Just a few small comments, this looks great!

A couple of big picture things:
- Can you convert the views to slim?  
- Let's add some tests for logging in and logging out.

**Comment by giladshanan** [comment]
@r3drighthand let me know if you have any questions

**Comment by r3drighthand** [comment]
@giladshanan 

Yeah, I started writing my view files in Slim but I don't have a linter or support installed for Slim on VSCode. I couldn't tell if my syntax was accurate because all text was white and it annoyed me. I'll go back and change the erb to Slim and fix any remaining single-quotes I find. 

I want to add a couple more views. Then I'll write some tests for logging in & out.

**Review by giladshanan — APPROVED** [review]
Ok sounds good. Should we add some cards to Trello so we don’t lose track of these ideas- Slim, double quotes, login/logout tests? Setting up the test environment should be its own PR, assuming you’ll be using rspec, circleCI, and possibly simplecov and rubocop as well. 

What do you think about using email_field_tag? If you don’t want to make the change now, we should add a card for email validation so we don’t forget about it later.

**Comment by r3drighthand** [comment]
@giladshanan 

I did some editing last night and changed all single to double quotes and swapped out the text_field_tag for email_field_tag. I'm rewriting my erb in Slim.

I agree that adding tests should happen in a separate branch for testing, as should my idea to add a basic landing page we can click around. I think the langing page should happen first, then the tests because right now we still have the generic You're On Rails! page and that doesn't help us at all. 

Should I go ahead and merge this PR, but continue working on my local setup-sorcery-pw-auth? It'll have erb > slim edits not included in this PR. I don't think it'll be n issue.

**Pull request merged**